File: estruturas/estruturasFlexiveis/exercicios/ListaDupla.java
package estruturasFlexiveis.exercicios;

public class ListaDupla {
    public static void main(String[] args) throws Exception {
        LinkedList listaDupla = new LinkedList();
        listaDupla.inserirInicio(2);
        listaDupla.inserirInicio(3);
        listaDupla.inserirInicio(4);
        listaDupla.mostrar();
        listaDupla.inserirFim(10);
        listaDupla.inserirFim(20);
        listaDupla.mostrar();
        listaDupla.removerInicio();
        listaDupla.mostrar();
        listaDupla.removerFim();
        listaDupla.mostrar();
        listaDupla.inserir(17, 2);
        listaDupla.mostrar();
        listaDupla.inserir(13, 2);
        listaDupla.mostrar();
        listaDupla.remover(1);
        listaDupla.mostrar();
    }
}

class CelulaDupla {
    int elemento;
    CelulaDupla ant, prox;

    CelulaDupla() {
        this(0);
    }

    CelulaDupla(int x) {
        elemento = x;
        ant = prox = null;
    }
}

class LinkedList {
    CelulaDupla primeiro, ultimo;

    LinkedList() {
        primeiro = new CelulaDupla();
        ultimo = primeiro;
    }

    void inserirInicio(int x) {
        CelulaDupla temp = new CelulaDupla(x);
        temp.prox = primeiro.prox;
        temp.ant = primeiro;
        primeiro.prox = temp;
        if (primeiro == ultimo) {
            ultimo = temp;
        } else {
            temp.prox.ant = temp;
        }
        temp = null;
    }

    void inserirFim(int x) {
        ultimo.prox = new CelulaDupla(x);
        ultimo.prox.ant = ultimo;
        ultimo = ultimo.prox;
    }

    int removerInicio() throws Exception {
        if (primeiro == ultimo) {
            throw new Exception("LISTA VAZIA");
        }
        CelulaDupla temp = primeiro.prox;
        int elemento = temp.elemento;
        primeiro.prox = temp.prox;
        temp.prox.ant = primeiro;
        temp.ant = temp.prox = null;
        temp = null;
        return elemento;
    }

    int removerFim() throws Exception {
        if (primeiro == ultimo) {
            throw new Exception("LISTA VAZIA");
        }
        int elemento = ultimo.elemento;
        CelulaDupla temp = ultimo.ant;
        temp.prox = ultimo.prox;
        ultimo.ant = null;
        ultimo = temp;
        temp = null;
        return elemento;
    }

    void mostrar() {
        System.out.println();
        CelulaDupla i;
        for (i = primeiro.prox; i != null; i = i.prox) {
            System.out.print(i.elemento + " ");
        }
    }

    void inserir(int x, int pos) throws Exception {
        int tam = tamanho();
        if (pos > tam || pos < 0) {
            throw new Exception("POSICAO INVALIDA");
        }
        if (pos == 0) {
            inserirInicio(x);
        } else if (pos == tam) {
            inserirFim(x);
        } else {
            CelulaDupla i = primeiro;
            for (int j = 0; j < pos; j++, i = i.prox)
                ;
            CelulaDupla temp = new CelulaDupla(x);
            temp.ant = i;
            temp.prox = i.prox;
            i.prox.ant = temp;
            i.prox = temp;
            temp = null;
        }
    }

    int remover(int pos) throws Exception {
        int tam = tamanho();
        if (pos > tam || pos < 0) {
            throw new Exception("POSICAO INVALIDA");
        }
        int elemento;
        if (pos == 0) {
            elemento = removerInicio();
        } else if (pos == tam - 1) {
            elemento = removerFim();
        } else {
            CelulaDupla i = primeiro;
            for (int j = 0; j < pos; j++, i = i.prox);
            CelulaDupla temp = i.prox;
            elemento = temp.elemento;
            i.prox = temp.prox;
            temp.prox.ant = i;
            temp.ant = temp.prox = null;
            temp = null;
        }
        return elemento;
    }

    int tamanho() {
        int cont = 0;
        for (CelulaDupla i = primeiro.prox; i != null; i = i.prox) {
            cont++;
        }
        return cont;
    }
}

File: estruturas/estruturasFlexiveis/lista/dupla/listaDupla.c
#include<stdio.h>
#include<stdlib.h>

typedef struct Celula {
    int elemento;
    struct Celula *prox;
    struct Celula *ant;
} Celula;

typedef struct Lista {
    struct Celula *primeiro;
    struct Celula *ultimo;
} Lista;

Celula* iniciarCelula() {
    Celula *celula = (Celula*)malloc(sizeof(Celula));
    celula->elemento = 0;
    celula->ant = NULL;
    celula->prox = NULL;
    return celula;
}

Lista* iniciarLista() {
    Lista *lista = (Lista*)malloc(sizeof(Lista));
    lista->primeiro = iniciarCelula();
    lista->ultimo = lista->primeiro;
    return lista;
}

void inserirInicio(Lista *lista, int x) {
    Celula *temp = iniciarCelula();
    temp->elemento = x;
    if (lista->primeiro->prox == NULL) {
        lista->primeiro->prox = temp;
        lista->ultimo = temp;
        temp->ant = lista->primeiro;
    } else {
        temp->prox = lista->primeiro->prox;
        temp->ant = lista->primeiro;
        lista->primeiro->prox->ant = temp;
        lista->primeiro->prox = temp;
    }
}

void inserirFim(Lista *lista, int x) {
    Celula *temp = iniciarCelula();
    temp->elemento = x;
    temp->ant = lista->ultimo;
    lista->ultimo->prox = temp;
    lista->ultimo = temp;
}

int removerInicio(Lista *lista) {
    if (lista->primeiro == lista->ultimo) {
        printf("LISTA VAZIA\n");
        exit(1);
    }
    Celula *temp = lista->primeiro->prox;
    int elemento = temp->elemento;
    lista->primeiro->prox = temp->prox;
    if (temp->prox != NULL) {
        temp->prox->ant = lista->primeiro;
    } else {
        lista->ultimo = lista->primeiro;
    }
    free(temp);
    return elemento;
}

int removerFim(Lista *lista) {
    if (lista->primeiro == lista->ultimo) {
        printf("LISTA VAZIA\n");
        exit(1);
    }
    Celula *temp = lista->ultimo;
    int elemento = temp->elemento;
    lista->ultimo = temp->ant;
    lista->ultimo->prox = NULL;
    free(temp);
    return elemento;
}

void mostrar(Lista *lista) {
    printf("\n");
    for (Celula *i = lista->primeiro->prox; i != NULL; i = i->prox) {
        printf("%d ", i->elemento);
    }
    printf("\n");
}

void inserir(Lista *lista, int pos, int x) {
    int tam = tamanho(lista);
    if (pos < 0 || pos > tam) {
        printf("POSICAO INVALIDA\n");
        exit(1);
    }
    if (pos == tam) {
        inserirFim(lista, x);
    } else if (pos == 0) {
        inserirInicio(lista, x);
    } else {
        Celula *i = lista->primeiro;
        for (int j = 0; j < pos; j++, i = i->prox);
        Celula *temp = iniciarCelula();
        temp->elemento = x;
        temp->ant = i;
        temp->prox = i->prox;
        i->prox->ant = temp;
        i->prox = temp;
    }
}

int remover(Lista *lista, int pos) {
    if (lista->primeiro == lista->ultimo) {
        printf("LISTA VAZIA\n");
        exit(1);
    }
    int tam = tamanho(lista);
    if (pos < 0 || pos >= tam) {
        printf("POSICAO INVALIDA\n");
        exit(1);
    }
    if (pos == 0) {
        return removerInicio(lista);
    } else if (pos == tam - 1) {
        return removerFim(lista);
    } else {
        Celula *i = lista->primeiro;
        for (int j = 0; j < pos; j++, i = i->prox);
        Celula *temp = i->prox;
        int elemento = temp->elemento;
        i->prox = temp->prox;
        temp->prox->ant = i;
        free(temp);
        return elemento;
    }
}

int tamanho(Lista *lista) {
    int cont = 0;
    for (Celula *i = lista->primeiro->prox; i != NULL; i = i->prox) {
        cont++;
    }
    return cont;
}

int main() {
    Lista *lista = iniciarLista();
    inserirInicio(lista, 10);
    inserirInicio(lista, 50);
    inserirInicio(lista, 30);
    inserirInicio(lista, 1);
    inserirInicio(lista, 90);
    inserirInicio(lista, 70);
    inserirInicio(lista, 6);
    mostrar(lista);
    inserirFim(lista, 20);
    mostrar(lista);
    removerInicio(lista);
    removerFim(lista);
    mostrar(lista);
    inserir(lista, 2, 17);
    mostrar(lista);
    remover(lista, 3);
    mostrar(lista);
}
